### PR TITLE
✨ Interchain Security Module

### PR DESCRIPTION
**Pull request by JordyRo1**

### CODE REVIEW

File: src/contracts/isms/multisig/messageid_multisig_ism.cairo [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            let mut i = 0;
+
+            // for each couple (sig_s, sig_r) extracted from the metadata
+            loop {
+                if (i == threshold) {
+                    break ();
+                }
+                let signature = get_signature_at(_metadata.clone(), i);
+
+                // we loop on the validators list public kew in order to find a match
```
Comment (EvolveArt) [inline src/contracts/isms/multisig/messageid_multisig_ism.cairo:57]:
```suggestion
                // we loop on the validators list public key in order to find a match
```

File: src/contracts/isms/multisig/multisig_ism.cairo [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    #[abi(embed_v0)]
+    impl OwnableImpl = OwnableComponent::OwnableImpl<ContractState>;
+    impl OwnableInternalImpl = OwnableComponent::InternalImpl<ContractState>;
+    impl UpgradeableInternalImpl = UpgradeableComponent::InternalImpl<ContractState>;
+    type Index = felt252;
+
+    #[derive(Serde, Copy, starknet::Store, Drop, PartialEq)]
+    pub struct ValidatorInformations {
+        pub address: EthAddress,
+        pub public_key: u256
```
Comment (EvolveArt) [inline src/contracts/isms/multisig/multisig_ism.cairo:21]: remove public key

File: src/contracts/isms/multisig/multisig_ism.cairo [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    component!(path: OwnableComponent, storage: ownable, event: OwnableEvent);
+    component!(path: UpgradeableComponent, storage: upgradeable, event: UpgradeableEvent);
+    #[abi(embed_v0)]
+    impl OwnableImpl = OwnableComponent::OwnableImpl<ContractState>;
+    impl OwnableInternalImpl = OwnableComponent::InternalImpl<ContractState>;
+    impl UpgradeableInternalImpl = UpgradeableComponent::InternalImpl<ContractState>;
+    type Index = felt252;
+
+    #[derive(Serde, Copy, starknet::Store, Drop, PartialEq)]
+    pub struct ValidatorInformations {
```
Comment (EvolveArt) [inline src/contracts/isms/multisig/multisig_ism.cairo:19]:
```suggestion
    pub struct ValidatorInformation {
```
typo

File: src/contracts/isms/multisig/validator_announce.cairo [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+            ref self: ContractState,
+            _validator: EthAddress,
+            _storage_location: felt252,
+            _signature: Bytes
+        ) -> bool {
+            let felt252_validator: felt252 = _validator.into();
+            let mut input: Array<u256> = array![
+                felt252_validator.into(), _storage_location.into(),
+            ];
+            let replay_id = keccak_hash(input.span());
```
Comment (EvolveArt) [inline src/contracts/isms/multisig/validator_announce.cairo:61]: why use keccack here ? this is just used internally by our contract right ?

File: src/contracts/isms/multisig/validator_announce.cairo [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+            let mut input: Array<u256> = array![
+                ETH_SIGNED_MESSAGE.into(), domain_hash.into(), _storage_location.into(),
+            ];
+            let hash = keccak_u256s_be_inputs(input.span());
+            reverse_endianness(hash)
+        }
+    }
+
+
+    fn convert_to_signature(_signature: Bytes) -> Signature {
```
Comment (EvolveArt) [inline src/contracts/isms/multisig/validator_announce.cairo:117]: write unit test for this

File: src/contracts/isms/multisig/validator_announce.cairo [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        let hash = keccak_u256s_be_inputs(_input);
+        reverse_endianness(hash)
+    }
+
+
+    fn domain_hash(self: @ContractState) -> u256 {
+        let mailboxclient = IMailboxClientDispatcher {
+            contract_address: self.mailboxclient.read()
+        };
+        let mailboxcient_address: felt252 = self.mailboxclient.read().try_into().unwrap();
```
Comment (EvolveArt) [inline src/contracts/isms/multisig/validator_announce.cairo:133]:
```suggestion
        let mailboxclient_address: felt252 = self.mailboxclient.read().try_into().unwrap();
```

File: src/contracts/isms/multisig/validator_announce.cairo [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    #[constructor]
+    fn constructor(ref self: ContractState, _mailbox_client: ContractAddress) {
+        self.mailboxclient.write(_mailbox_client);
+    }
+
+    #[abi(embed_v0)]
+    impl IValidatorAnnonceImpl of IValidatorAnnounce<ContractState> {
+        fn announce(
+            ref self: ContractState,
+            _validator: EthAddress,
```
Comment (EvolveArt) [inline src/contracts/isms/multisig/validator_announce.cairo:53]: what if validator is the null address ?

File: src/contracts/isms/multisig/multisig_ism.cairo [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+            loop {
+                if (cur_idx == _validators.len()) {
+                    break ();
+                }
+                let validator = *_validators.at(cur_idx);
+                assert(
+                    validator.address != 0.try_into().unwrap(),
+                    Errors::VALIDATOR_ADDRESS_CANNOT_BE_NULL
+                );
+                // assert(validator.public_key != 0, Errors::VALIDATOR_PUBLIC_KEY_CANNOT_BE_NULL);
```
Comment (EvolveArt) [inline src/contracts/isms/multisig/multisig_ism.cairo:77]: remove comment

File: src/contracts/isms/multisig/multisig_ism.cairo [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+    #[constructor]
+    fn constructor(ref self: ContractState, _owner: ContractAddress) {
+        self.ownable.initializer(_owner);
+    }
+
+
+    mod Errors {
+        pub const VALIDATOR_ADDRESS_CANNOT_BE_NULL: felt252 = 'Validator address cannot be 0';
+        pub const VALIDATOR_PUBLIC_KEY_CANNOT_BE_NULL: felt252 = 'Validator pk cannot be 0';
```
Comment (EvolveArt) [inline src/contracts/isms/multisig/multisig_ism.cairo:51]: remove

**Review by EvolveArt — CHANGES_REQUESTED** [review]

File: src/contracts/mocks/ism.cairo [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    impl IMessageidMultisigIsmImpl of IInterchainSecurityModule<ContractState> {
+        fn module_type(self: @ContractState) -> ModuleType {
+            ModuleType::MESSAGE_ID_MULTISIG(starknet::get_contract_address())
+        }
+
+        fn verify(
+            self: @ContractState,
+            _metadata: Bytes,
+            _message: Message,
+            _validator_configuration: ContractAddress
```
Comment (EvolveArt) [inline src/contracts/mocks/ism.cairo:23]: why this ?

File: src/contracts/isms/multisig/messageid_multisig_ism.cairo [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+    impl IMessageidMultisigIsmImpl of IInterchainSecurityModule<ContractState> {
+        fn module_type(self: @ContractState) -> ModuleType {
+            ModuleType::MESSAGE_ID_MULTISIG(starknet::get_contract_address())
+        }
+
+        fn verify(
+            self: @ContractState,
+            _metadata: Bytes,
+            _message: Message,
+            _validator_configuration: ContractAddress,
```
Comment (EvolveArt) [inline src/contracts/isms/multisig/messageid_multisig_ism.cairo:35]: afaik this arg is not in the spec

File: src/contracts/mocks/ism.cairo (hunk 10 above)
Comment (JordyRo1) [reply src/contracts/mocks/ism.cairo:23]: _validator_configuration refers to the contract defining the set of validators and the threshold conditions. The user can simply mock this contract and specify conditions on validator selections. https://docs.hyperlane.xyz/docs/reference/ISM/multisig-ISM-interface

Comment (EvolveArt) [reply src/contracts/mocks/ism.cairo:23]: yes. but its not the spec, so we should remove this arg cc https://docs.hyperlane.xyz/docs/reference/ISM/specify-your-ISM

**Review by EvolveArt — APPROVED** [review]

**Pull request merged**